pkg/codegen/schema/loader.go

```diff
@@ -51,12 +56,66 @@ func (l *pluginLoader) ensurePlugin(pkg string, version *semver.Version) error {
 		Name:    pkg,
 		Version: version,
 	}
+
+	tryDownload := func(dst io.WriteCloser) error {
+		defer dst.Close()
+		tarball, expectedByteCount, err := pkgPlugin.Download()
+		if err != nil {
+			return err
+		}
+		defer tarball.Close()
+		copiedByteCount, err := io.Copy(dst, tarball)
+		if err != nil {
+			return err
+		}
+		if copiedByteCount != expectedByteCount {
+			return fmt.Errorf("Expected %d bytes but copied %d when downloading plugin %s",
+				expectedByteCount, copiedByteCount, pkgPlugin)
+		}
+		return nil
+	}
+
+	tryDownloadToFile := func() (string, error) {
+		file, err := ioutil.TempFile("" /* default temp dir */, "pulumi-plugin-tar")
+		if err != nil {
+			return "", err
+		}
+		err = tryDownload(file)
+		if err != nil {
+			err2 := os.Remove(file.Name())
+			if err2 != nil {
+				return "", fmt.Errorf("Error while removing tempfile: %v. Context: %w", err2, err)
+			}
+			return "", err
+		}
+		return file.Name(), nil
+	}
+
+	downloadToFileWithRetry := func() (string, error) {
+		for attempt := 0; ; attempt++ {
+			tempFile, err := tryDownloadToFile()
+			if err == nil {
+				return tempFile, nil
+			}
+
+			if err != nil && attempt >= 5 {
+				return tempFile, err
+			}
+			time.Sleep(100 * time.Microsecond)
+		}
+	}
+
 	if !workspace.HasPlugin(pkgPlugin) {
-		tarball, _, err := pkgPlugin.Download()
+		tarball, err := downloadToFileWithRetry()
 		if err != nil {
 			return errors.Wrapf(err, "failed to download plugin: %s", pkgPlugin)
 		}
-		if err := pkgPlugin.Install(tarball); err != nil {
+		defer os.Remove(tarball)
+		reader, err := os.Open(tarball)
+		if err != nil {
+			return errors.Wrapf(err, "failed to open downloaded plugin: %s", pkgPlugin)
+		}
+		if err := pkgPlugin.Install(reader); err != nil {
 			return errors.Wrapf(err, "failed to install plugin %s", pkgPlugin)
```

Comment:
Just for clarification, are we downloading to a file s.t. we can retry just the download rather than the whole call to `Install`?

Reply:
Hm I wasn't sure we could retry Install but we could do it that way too guess, it may be idempotent. Conceptually it feels slightly cleaner to attribute the error to Download since it's a network error. But there's a cost to the temp file.  Let's try this but if there's reliability or disk space issues around temp files we can switch to retrying install.

Reply:
SGTM